### ISSUE TITLE
code generated fails on dialyzer

### ISSUE THREAD

**Issue by Licenser** [lrascao/rebar3_gpb_plugin#51]
the code gpb generates causes dialyzer error:
```
apps/ddb_proxy/src/remote_pb.erl
  38: Type specification remote_pb:encode_msg(_) -> binary() is a supertype of the success typing: remote_pb:encode_msg(#'WriteRequest'{timeseries::[#'TimeSeries'{labels::[any()],samples::[any()]}]} | #'LabelPair'{name::'undefined' | binary() | maybe_improper_list(binary() | maybe_improper_list(any(),binary() | []) | byte(),binary() | []),value::'undefined' | binary() | maybe_improper_list(binary() | maybe_improper_list(any(),binary() | []) | byte(),binary() | [])} | #'Sample'{value::'-infinity' | 'infinity' | 'nan' | 'undefined' | number(),timestamp_ms::'undefined' | integer()} | #'TimeSeries'{labels::[#'LabelPair'{name::'undefined' | binary() | maybe_improper_list(any(),binary() | []),value::'undefined' | binary() | maybe_improper_list(any(),binary() | [])}],samples::[#'Sample'{value::'-infinity' | 'infinity' | 'nan' | 'undefined' | number(),timestamp_ms::'undefined' | integer()}]}) -> binary()
  42: Type specification remote_pb:encode_msg(_,[any()]) -> binary() is a supertype of the success typing: remote_pb:encode_msg(#'WriteRequest'{timeseries::[#'TimeSeries'{labels::[any()],samples::[any()]}]} | #'LabelPair'{name::'undefined' | binary() | maybe_improper_list(binary() | maybe_improper_list(any(),binary() | []) | byte(),binary() | []),value::'undefined' | binary() | maybe_improper_list(binary() | maybe_improper_list(any(),binary() | []) | byte(),binary() | [])} | #'Sample'{value::'-infinity' | 'infinity' | 'nan' | 'undefined' | number(),timestamp_ms::'undefined' | integer()} | #'TimeSeries'{labels::[#'LabelPair'{name::'undefined' | binary() | maybe_improper_list(any(),binary() | []),value::'undefined' | binary() | maybe_improper_list(any(),binary() | [])}],samples::[#'Sample'{value::'-infinity' | 'infinity' | 'nan' | 'undefined' | number(),timestamp_ms::'undefined' | integer()}]},[any()]) -> binary()
```

**Comment by lrascao** [comment]
@tomas-abrahamsson is this expected?

**Comment by tomas-abrahamsson** [comment]
Generally, no. I've tried to make both gpb itself and also the generated code it generates run clean in dialyzer. But dialyzing everything has been a bit slow, so dialyzing has never been part of any regression tests. I've done it manually from time to time and there have been a few regressions. There may be issues though, of course. @Licenser, could you share the .proto file(s) you're compiling, the gpb version and gpb options? (or possibly distilled .proto files if you can't share them)

As a side note, I should say also that with Erlang 20, I believe, dialyzing appears to go much faster, making it possible to dialyze also large and many (generated) files quickly, so it's beginning to make sense to regression test this.

**Comment by Licenser** [comment]
Hi tomas, thank you for the quick reply :D.

And sorry, I should have included that in the beginning.

* The proto file is this: https://github.com/dalmatinerdb/ddb_proxy/blob/test/apps/ddb_proxy/proto/remote.proto (Prometheus remote writer protocol)
* The gdp plugin vers: `1.11.8` (latest hex)
* The config is the default one (https://github.com/dalmatinerdb/ddb_proxy/blob/test/rebar.config)

**Comment by tomas-abrahamsson** [comment]
Thanks, I will take a closer look, hopefully tomorrow.

**Comment by Licenser** [comment]
Thank you so much! :) no hurry I just wanted to ticket it.

**Comment by tomas-abrahamsson** [comment]
Ah, `-Wunderspecs`, that's why I hadn't seen it. It looks like it will be relatively easy to fix.

**Comment by tomas-abrahamsson** [comment]
Thanks to @moxley and @kbaird fixing this in tomas-abrahamsson/gpb#118! But I experimented a bit further, and found in the general case it will be very difficult producing code that always passes `dialyzer -Wunderspecs`. Here is an example. Let's say we have `bool2.proto`:
```protobuf
    message m1 {
      required bool f1 = 1;
    }
```
When compiled with `-type` `-maps` and `-maps_unset_optional omitted`, gpb generates code like this:
```erlang
-type m1() ::
      #{f1                      := boolean() | 0 | 1 % = 1
       }.
-export_type(['m1'/0]).

-spec encode_msg(m1(),m1) -> binary().
encode_msg(Msg, MsgName) ->
    encode_msg(Msg, MsgName, []).
```
While this may look good, `dialyzer --plt .gpb-20.1.1.plt -Wunderspecs bool2.beam` produces a 
warning (I've reformatted it to make it a bit easier to read)
```
Type specification
    
         bool2:encode_msg(m1(),'m1') -> binary()
    
      is a supertype of the success typing:
    
         bool2:encode_msg(#{'f1':=boolean() | 0 | 1},'m1') -> <<_:8,_:_*8>>
```
My interpretation is that dialyzer says that the return type, `binary()`, can be further narrowed down to `<<_:8,_:_*8>>` (the arguments' types matches fairly well as far as I can see, so it must be the return type that dialyzer has opinions about) In this particular case, future versions of dialyzer might even be smart enough to conclude that the return type could be narrowed down to `<<8, 0 | 1>>`.

But in my opinion, `binary()` is the return type I'd prefer (not only is it easier to read, but also,  `<<_:8,_:_*8>> and even `<<8,0|1>>` are a bit too precise, and will be difficult to accurately generate in the general case) and this in turn would mean that there can always be cases where `dialyzer -Wunderspecs` can produce warnings. Thus, this report is _Won't Fix_, but I will merge the tomas-abrahamsson/gpb#118, because although it cannot solve the general case, the generated code becomes better. (sorry this comment became a bit long)

**Comment by lrascao** [comment]
don't know if it's the right approach here, but in the past i have got rid of these kind of warnings but doing `<< (SomeBinary/binary) >>.´, not elegant but it shuts up dialyzer

**Comment by tomas-abrahamsson** [comment]
I'm unsure, but will that not actually do an extra (unnecessary) copy of SomeBinary?

**Comment by tomas-abrahamsson** [comment]
Thanks to @moxley and @kbaird! Their fix is included in 3.28.0 which I just pushed.  You might want to include `type_specs` in `gpb_opts`, to make dialyzer more happy with the generated code. For me, remote.proto now dialyzes cleanly.

I am not allowed by github to close this ticket, but if you're ok with 3.28.0, you might perhaps do that?

**Comment by lrascao** [comment]
closing, thanks all
